lib/statespace_sse.h

```diff
@@ -0,0 +1,214 @@
+      return buffer[0] + buffer[1] + buffer[2] + buffer[3];
+    };
+
+    return ParallelFor::RunReduce(
+        Base::num_threads_, Base::raw_size_ / 8, f, Op(), state);
+  }
+
+  std::complex<double> InnerProduct(
+      const State& state1, const State& state2) const {
+    using Op = std::plus<std::complex<double>>;
+
+    auto f = [](unsigned n, unsigned m, uint64_t i, const State& state1,
+              const State& state2) -> std::complex<double> {
+      __m128 re1 = _mm_load_ps(state1.get() + 8 * i);
+      __m128 im1 = _mm_load_ps(state1.get() + 8 * i + 4);
+      __m128 re2 = _mm_load_ps(state2.get() + 8 * i);
+      __m128 im2 = _mm_load_ps(state2.get() + 8 * i + 4);
+
+      __m128 ip_re = _mm_add_ps(_mm_mul_ps(re1, re2), _mm_mul_ps(im1, im2));
+      __m128 ip_im = _mm_sub_ps(_mm_mul_ps(re1, im2), _mm_mul_ps(im1, re2));
+
+      float bre[4];
+      float bim[4];
+      _mm_storeu_ps(bre, ip_re);
+      _mm_storeu_ps(bim, ip_im);
+
+      double re = bre[0] + bre[1] + bre[2] + bre[3];
+      double im = bim[0] + bim[1] + bim[2] + bim[3];
+
+      return std::complex<double>{re, im};
+    };
+
+    return ParallelFor::RunReduce(
+        Base::num_threads_, Base::raw_size_ / 8, f, Op(), state1, state2);
+  }
+
+  double RealInnerProduct(const State& state1, const State& state2) const {
+    using Op = std::plus<double>;
+
+    auto f = [](unsigned n, unsigned m, uint64_t i, const State& state1,
+              const State& state2) -> double {
+      __m128 re1 = _mm_load_ps(state1.get() + 8 * i);
+      __m128 im1 = _mm_load_ps(state1.get() + 8 * i + 4);
+      __m128 re2 = _mm_load_ps(state2.get() + 8 * i);
+      __m128 im2 = _mm_load_ps(state2.get() + 8 * i + 4);
+
+      __m128 ip_re = _mm_add_ps(_mm_mul_ps(re1, re2), _mm_mul_ps(im1, im2));
+
+      float bre[4];
+      _mm_storeu_ps(bre, ip_re);
+
+      return bre[0] + bre[1] + bre[2] + bre[3];
+    };
+
+    return ParallelFor::RunReduce(
+        Base::num_threads_, Base::raw_size_ / 8, f, Op(), state1, state2);
+  }
+
+  template <typename DistrRealType = double>
+  std::vector<uint64_t> Sample(
+      const State& state, uint64_t num_samples, unsigned seed) const {
+    std::vector<uint64_t> bitstrings;
+
+    if (num_samples > 0) {
+      double norm = 0;
+      uint64_t size = Base::raw_size_ / 8;
+      const float* v = state.get();
+
+      for (uint64_t k = 0; k < size; ++k) {
+        for (unsigned j = 0; j < 4; ++j) {
+          auto re = v[8 * k + j];
+          auto im = v[8 * k + 4 + j];
+          norm += re * re + im * im;
+        }
+      }
+
+      auto rs = GenerateRandomValues<DistrRealType>(num_samples, seed, norm);
+
+      uint64_t m = 0;
+      double csum = 0;
+      bitstrings.reserve(num_samples);
+
+      for (uint64_t k = 0; k < size; ++k) {
+        for (unsigned j = 0; j < 4; ++j) {
+          auto re = v[8 * k + j];
+          auto im = v[8 * k + 4 + j];
+          csum += re * re + im * im;
+          while (rs[m] < csum && m < num_samples) {
```

Comment:
Think the order needs to be swapped here like in #80

Reply:
This is intentional. `while (rs[m] < csum && m < num_samples)` is slightly faster than `while (m < num_samples && rs[m] < csum)`. Note that there is no out of boundary memory access as I allocate one extra array element in `GenerateRandomValues` in `util.h`.

Reply:
See #80 - to fix the sanitizer issue, we should populate that extra element.